File: app/src/main/java/za/co/woolworths/financial/services/android/util/expand/ParentSubCategoryViewHolder.java
package za.co.woolworths.financial.services.android.util.expand;

import android.view.View;
import android.view.animation.RotateAnimation;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;

import com.awfs.coordination.R;

import za.co.woolworths.financial.services.android.ui.views.WTextView;

public class ParentSubCategoryViewHolder extends ParentViewHolder {

	private static final float INITIAL_POSITION = 0.0f;
	private static final float ROTATED_POSITION = 180f;
	private final ImageView mArrowExpandImageView;
	private final ProgressBar pbLoadChildItem;
	private WTextView tvSubCategoryName;
	private LinearLayout llLoadChild;

	public ParentSubCategoryViewHolder(View itemView) {
		super(itemView);
		tvSubCategoryName = itemView.findViewById(R.id.carbon_groupText);
		llLoadChild = itemView.findViewById(R.id.llLoadChild);
		mArrowExpandImageView = itemView.findViewById(R.id.carbon_groupExpandedIndicator);
		pbLoadChildItem = itemView.findViewById(R.id.pbLoadChildItem);
	}

	public void bind(SubCategoryModel subCategoryModel) {
		tvSubCategoryName.setText(subCategoryModel.getName());
		arrowVisibility(subCategoryModel);
		retrieveChildVisibility(subCategoryModel);
	}

	private void arrowVisibility(SubCategoryModel subCategoryModel) {
		llLoadChild.setVisibility(subCategoryModel.getSubCategory().hasChildren ? View.VISIBLE : View.INVISIBLE);

	}

	public void retrieveChildVisibility(SubCategoryModel subCategoryModel) {
		boolean progressBarIsVisible = subCategoryModel.getSubCategory().singleProductItemIsLoading;
		pbLoadChildItem.setVisibility(progressBarIsVisible ? View.VISIBLE : View.GONE);
		mArrowExpandImageView.setVisibility(progressBarIsVisible ? View.GONE : View.VISIBLE);
	}

	public void retrieveChildVisibility(boolean visible) {
		pbLoadChildItem.setVisibility(visible ? View.VISIBLE : View.GONE);
		mArrowExpandImageView.setVisibility(visible ? View.GONE : View.VISIBLE);
	}

	@Override
	public void setExpanded(boolean expanded) {
		super.setExpanded(expanded);
		if (expanded) {
			mArrowExpandImageView.setRotation(ROTATED_POSITION);
		} else {
			mArrowExpandImageView.setRotation(INITIAL_POSITION);
		}
	}

	@Override
	public void onExpansionToggled(boolean expanded) {
		super.onExpansionToggled(expanded);

		RotateAnimation rotateAnimation;
		if (expanded) { // rotate clockwise
			rotateAnimation = new RotateAnimation(ROTATED_POSITION,
					INITIAL_POSITION,
					RotateAnimation.RELATIVE_TO_SELF, 0.5f,
					RotateAnimation.RELATIVE_TO_SELF, 0.5f);
		} else { // rotate counterclockwise
			rotateAnimation = new RotateAnimation(-1 * ROTATED_POSITION,
					INITIAL_POSITION,
					RotateAnimation.RELATIVE_TO_SELF, 0.5f,
					RotateAnimation.RELATIVE_TO_SELF, 0.5f);
		}

		rotateAnimation.setDuration(200);
		rotateAnimation.setFillAfter(true);
		mArrowExpandImageView.startAnimation(rotateAnimation);

	}
}

File: app/src/main/java/za/co/woolworths/financial/services/android/util/expand/ParentSubCategoryViewHolder.kt
package za.co.woolworths.financial.services.android.util.expand

import android.view.View
import android.view.animation.RotateAnimation
import com.awfs.coordination.databinding.SubCategoryParentViewBinding
import za.co.woolworths.financial.services.android.util.ImageManager.Companion.setPictureWithoutPlaceHolder

class ParentSubCategoryViewHolder(val itemBinding: SubCategoryParentViewBinding) : ParentViewHolder(itemBinding.root) {

    fun bind(subCategoryModel: SubCategoryModel) {
        itemBinding.carbonGroupText?.setText(subCategoryModel.name)
        arrowVisibility(subCategoryModel)
        retrieveChildVisibility(subCategoryModel)
        setNewBadgeImage(subCategoryModel)
    }

    private fun setNewBadgeImage(subCategoryModel: SubCategoryModel) {
        if (subCategoryModel.imageUrl != null) {
            itemBinding.imgBadge?.visibility = View.VISIBLE
            setPictureWithoutPlaceHolder(itemBinding.imgBadge, subCategoryModel.imageUrl)
        } else {
            itemBinding.imgBadge?.visibility = View.GONE
        }
    }

    private fun arrowVisibility(subCategoryModel: SubCategoryModel) {
        itemBinding.llLoadChild?.visibility =
            if (subCategoryModel.subCategory.hasChildren) View.VISIBLE else View.INVISIBLE
    }

    private fun retrieveChildVisibility(subCategoryModel: SubCategoryModel) {
        val progressBarIsVisible = subCategoryModel.subCategory.singleProductItemIsLoading
        itemBinding.pbLoadChildItem?.visibility = if (progressBarIsVisible) View.VISIBLE else View.GONE
        itemBinding.carbonGroupExpandedIndicator?.visibility =
            if (progressBarIsVisible) View.GONE else View.VISIBLE
    }

    fun retrieveChildVisibility(visible: Boolean) {
        itemBinding.pbLoadChildItem?.visibility = if (visible) View.VISIBLE else View.GONE
        itemBinding.carbonGroupExpandedIndicator?.visibility = if (visible) View.GONE else View.VISIBLE
    }

    override fun setExpanded(expanded: Boolean) {
        super.isRowExpanded = expanded
        if (expanded) {
            itemBinding.carbonGroupExpandedIndicator?.rotation = ROTATED_POSITION
        } else {
            itemBinding.carbonGroupExpandedIndicator?.rotation = INITIAL_POSITION
        }
    }

    override fun onExpansionToggled(expanded: Boolean) {
        super.onExpansionToggled(expanded)
        val rotateAnimation: RotateAnimation = if (expanded) { // rotate clockwise
            RotateAnimation(ROTATED_POSITION,
                INITIAL_POSITION,
                RotateAnimation.RELATIVE_TO_SELF, 0.5f,
                RotateAnimation.RELATIVE_TO_SELF, 0.5f)
        } else { // rotate counterclockwise
            RotateAnimation(-1 * ROTATED_POSITION,
                INITIAL_POSITION,
                RotateAnimation.RELATIVE_TO_SELF, 0.5f,
                RotateAnimation.RELATIVE_TO_SELF, 0.5f)
        }
        rotateAnimation.duration = 200
        rotateAnimation.fillAfter = true
        itemBinding.carbonGroupExpandedIndicator?.startAnimation(rotateAnimation)
    }

    companion object {
        private const val INITIAL_POSITION = 0.0f
        private const val ROTATED_POSITION = 180f
    }

}